### ISSUE TITLE
JCEF preview never loads when using custom JavaScript in preview

### ISSUE THREAD

**Issue by ari** [asciidoctor/asciidoctor-intellij-plugin#567]
When trying to preview in JCEF mode I get this console error:
```
Refused to load the script 'file:///jbcefbrowser/js/jquery.js' because it violates the following Content Security Policy directive: "script-src 'unsafe-inline' 'unsafe-eval' http://localhost:63342/ead61b63-b0a6-4ff2-a49a-86be75ccfd1a/scripts/". Note that 'script-src-elem' was not explicitly set, so 'script-src' is used as a fallback.
1494964479#url=jar:f….class@2004291123:1 Refused to load the script 'file:///jbcefbrowser/js/bootstrap-scrollspy.js' because it violates the following Content Security Policy directive: "script-src 'unsafe-inline' 'unsafe-eval' http://localhost:63342/ead61b63-b0a6-4ff2-a49a-86be75ccfd1a/scripts/". Note that 'script-src-elem' was not explicitly set, so 'script-src' is used as a fallback.
1494964479#url=jar:f….class@2004291123:1 Refused to load the script 'file:///jbcefbrowser/js/lazy-loading-images.js' because it violates the following Content Security Policy directive: "script-src 'unsafe-inline' 'unsafe-eval' http://localhost:63342/ead61b63-b0a6-4ff2-a49a-86be75ccfd1a/scripts/". Note that 'script-src-elem' was not explicitly set, so 'script-src' is used as a fallback.
1494964479#url=jar:f….class@2004291123:1 Refused to load the script 'file:///jbcefbrowser/js/documentation.js' because it violates the following Content Security Policy directive: "script-src 'unsafe-inline' 'unsafe-eval' http://localhost:63342/ead61b63-b0a6-4ff2-a49a-86be75ccfd1a/scripts/". Note that 'script-src-elem' was not explicitly set, so 'script-src' is used as a fallback.
```

AsciiDoc Plugin Version: 0.31.33

IntelliJ IDEA 2020.2.3 (Ultimate Edition)
Build #IU-202.7660.26, built on October 6, 2020
Runtime version: 11.0.8+10-b944.34 x86_64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
macOS 10.15.6
GC: ParNew, ConcurrentMarkSweep
Memory: 2014M
Cores: 4

**Comment by ahus1** [comment]
Hello @ari, thanks for reporting this. I assume you're using some custom JavaScript in the preview? Can you please tell me a little bit more about your setup? 

The JCEF preview implements a content security policy (CSP); to make custom JavaScript work it probably need to make it either less restrictive or more configurable. An example will help me to test it. 

As the links are relative to the preview, it might be tricky to load custom JavaScript for the preview - after fixing CSP it might end up with a 404. Again, knowing more about your setup and what you want to achieve will help me to find a solution.

**Comment by ahus1** [comment]
Hi @ari - I looked into the repo and found the commit that changed it: The loading image is now pure HTML in the footer. The PR https://github.com/ishgroup/onCourseDocs/pull/2 will now only include the header, no longer the footer. This avoids the loading panel in the preview.

Still, the JavaScript files will not load in the preview as absolute paths in `docinfo` are not supported yet. 

Please let me know if this work for you.

**Comment by ari** [comment]
Hi @ahus1 Thanks so much! You've gone above and beyond to discover the cause here.

Do you mean that "related paths" aren't supported? I can't think of a solid workaround for this, since the location of the manual on the final destination path ( https://www.ish.com.au/onCourse/doc/manual/ ) is several folders deep. Unless you allow .asciidoctorconfig to specify a list of files to copy into a temp directory for rendering?

**Comment by ahus1** [comment]
Hi @ari - yes, "related paths" is probably the term, as a non-native speaker it's sometime hard for me. 

Anything that has a protocol header will work (like `https://fonts.googleapis.com/...`), everything that doesn't will not work. Workarounds would be to either inline the content of the file into docinfo.html, or to reference the files on the online site using a full URI (`https://www.ish.com.au/...`). For now, I would rather leave it as is. Some of the JavaScript might conflict with the incremental update of the preview.

A future version might parse the docinfo.html and rewrite relative URIs to point to the right locations and make them resolvable. Until that happens it is not supported. 

I [updated the docs](https://intellij-asciidoc-plugin.ahus1.de/docs/users-guide/features/advanced/stylesheets.html) to describe that this doesn't work yet. I'd be happy if you could do a review. 

Please let me know if the current functionality is sufficient for your current talk at hand.

**Comment by ari** [comment]
Of course I meant to write "relative path"! Sorry. Learning another language is hard enough without getting wrong hints.

Yes, the manual looks good and is clear. My particular problem is solved, but I still think some sort of "copy resources to local preview folder" idea might be useful for people who want to host fonts, etc locally inside their project.

I'll close this task. Thanks again for all the help.

**Comment by ahus1** [comment]
Hi Ari, I agree that local content referenced in docinfo should be supported. Having it buried in the docs as "not supported" doesn't help much. I opened issue #569 to follow up on this. 

The other restriction regarding attribute resolution in docinfo header/footer will be supported in the next release.

**Comment by ahus1** [comment]
Attribute resolution in docinfo header/footer is available as part or pre-release 0.31.35. 

The pre-release of the plugin is available from [GitHub releases](https://github.com/asciidoctor/asciidoctor-intellij-plugin/releases) and the [IntelliJ AsciiDoc EAP repository](https://intellij-asciidoc-plugin.ahus1.de/docs/contributors-guide/user/add-eap-repository-to-ide.html).
